lib/SIL/Utils/OwnershipUtils.cpp

```diff
@@ -1610,55 +1707,105 @@ bool ForwardingOperand::visitForwardedValues(
+      if (op->getOperandOwnership() == OperandOwnership::Reborrow) {
+        worklist.insert(std::make_tuple(PhiOperand(op), baseValue));
+      }
+      return true;
+    });
+  };
+
+  // Initialize the worklist.
+  collectReborrows(borrowInst, borrowInst->getOperand());
 
-  // Size of worklist changes in this loop
+  // For every (reborrow, base value) pair in the worklist:
+  // - Find phi value and new base value
+  // - Call the visitor on the phi value and new base value pair
+  // - Populate the worklist with pairs of reborrows of phi value and the new
+  // base.
   for (unsigned idx = 0; idx < worklist.size(); idx++) {
-    Operand *reborrowOp;
-    SILValue baseValue;
-    std::tie(reborrowOp, baseValue) = worklist[idx];
-
-    BorrowingOperand borrowingOperand(reborrowOp);
-    assert(borrowingOperand.isReborrow());
-
-    auto *branchInst = cast<BranchInst>(reborrowOp->getUser());
-    auto *succBlock = branchInst->getDestBB();
-    auto *phiArg = cast<SILPhiArgument>(
-        succBlock->getArgument(reborrowOp->getOperandNumber()));
-
-    SILValue newBaseVal = baseValue;
-    // If the previous base value was also passed as a phi arg, that will be
-    // the new base value.
-    for (auto *arg : succBlock->getArguments()) {
-      if (arg->getIncomingPhiValue(branchInst->getParent()) == baseValue) {
-        newBaseVal = arg;
-        break;
+    PhiOperand phiOp;
+    SILValue currentBaseValue;
+    std::tie(phiOp, currentBaseValue) = worklist[idx];
+
+    auto *phiValue = phiOp.getValue();
+    SILValue newBaseValue = currentBaseValue;
+
+    // If the previous base value was also passed as a branch arg along with the
+    // reborrow, its phi value will be the new base value.
+    for (auto &op : phiOp.getBranch()->getAllOperands()) {
+      PhiOperand otherPhiOp(&op);
+      if (otherPhiOp.getSource() != currentBaseValue) {
+        continue;
       }
+      newBaseValue = otherPhiOp.getValue();
     }
 
     // Call the visitor function
-    visitReborrowBaseValuePair(phiArg, newBaseVal);
+    visitReborrowPhiBaseValuePair(phiValue, newBaseValue);
 
-    BorrowedValue scopedValue(phiArg);
-    scopedValue.visitLocalScopeEndingUses([&](Operand *op) {
-      if (op->getOperandOwnership() == OperandOwnership::Reborrow) {
-        worklist.insert(std::make_tuple(op, newBaseVal));
+    collectReborrows(phiValue, newBaseValue);
+  }
+}
+
+void swift::visitExtendedGuaranteedForwardingPhiBaseValuePairs(
+    BorrowedValue borrow, function_ref<void(SILPhiArgument *, SILValue)>
+                              visitGuaranteedForwardingPhiBaseValuePair) {
+  assert(borrow.kind == BorrowedValueKind::BeginBorrow ||
+         borrow.kind == BorrowedValueKind::LoadBorrow);
+  // worklist stores (guaranteed forwarding phi, base value) pairs.
+  // We need a SetVector to make sure we don't revisit the same pair again.
+  SmallSetVector<std::tuple<PhiOperand, SILValue>, 4> worklist;
```

Comment:
We can have different base values on different paths:
```
sil [ossa] @test_forwarded_separate_base_values : $@convention(thin) (@owned Wrapper1) -> () {
bb0(%0 : @owned $Wrapper1):
  %outer = begin_borrow %0 : $Wrapper1
  cond_br undef, bb1, bb2 

bb1:
  %unrelated = begin_borrow %0 : $Wrapper1
  %ex1 = struct_extract %outer : $Wrapper1, #Wrapper1.val1
  br bb3(%ex1 : $Wrapper2, %unrelated : $Wrapper1)

bb2:
  %inner = begin_borrow %0 : $Wrapper1
  %ex2 = struct_extract %inner : $Wrapper1, #Wrapper1.val1
  br bb3(%ex2 : $Wrapper2, %inner : $Wrapper1)

bb3(%phi1 : @guaranteed $Wrapper2, %phi2 : @guaranteed $Wrapper1):
  %f = function_ref @use_wrapper2 : $@convention(thin) (@guaranteed Wrapper2) -> ()
  apply %f(%phi1) : $@convention(thin) (@guaranteed Wrapper2) -> ()
  end_borrow %phi2 : $Wrapper1
  end_borrow %outer : $Wrapper1
  destroy_value %0 : $Wrapper1
  %9999 = tuple()
  return %9999 : $() 
} 
```
Here %phi1 has %phi2 as base value on one path and %outer as base value in another.

Reply:
Comment long term, would like a "single-base" invariant!

Reply:
Instead of saying we want a "single-base" invariant. Please add the .sil above as a basic OSSA verification test case. Also please add the same test where the base is an owned value. That you can refer to the fact that a guaranteed phi can have multiple bases in a comment here, along with a stripped down version of the SIL. This comment and example should also appear at the top of the reborrow verifier!

Reply:
Done